### ISSUE TITLE
Refactor enums into (typed) enum classes

### ISSUE THREAD

**Issue by scarf005** [cataclysmbnteam/Cataclysm-BN#1527]
# Is your feature request related to a problem? Please describe.
Currently codebase uses mixed form of enum classes and enums, potentially harming readability, code safety, and name reusability.

# Describe the solution you'd like
### Migrate previous enum declarations into enum class.
```c
enum color {red, green, blue};
enum status {red, green, blue}; // Causes error
```
```c++
enum class color { red, green, blue };
enum class status { red, green, blue }; // OK
// use as color::red, status::red
```
### Use Typed enums
```c++
#include <iostream>

using namespace std;

enum class foo: int { a = 1000, b = 'b', c = 'c' };

int main() {
    char x = foo::a;
    return 0;
}
/*
main.cpp:8:19: error: cannot convert ‘foo’ to ‘char’ in initialization
    8 |     char x = foo::a;
      |              ~~~~~^
      |                   |
      |                   foo
*/
```
It also works for plain old enums.

# Describe alternatives you've considered
Work on other things, however better code readability and safety is nice.

# Additional context
- [cppreference.com page about enumeration declaration](https://en.cppreference.com/w/cpp/language/enum)
- [answer on stack overflow why enum class is preferred over plain enum](https://stackoverflow.com/a/18335862/13503626)

**Comment by Coolthulhu** [comment]
Yeah, a lot of the code still uses legacy C enums. In some cases, they are used as integers (nothing that an explicit `static_cast<size_t>` wouldn't solve) or bitflags (we have a macro to define boolean AND/OR/etc. operators on an `enum class` type), in other cases they are just plain named constants.

**Comment by scarf005** [comment]
could blindly changing them into enum class cause possible errors? i think it may be possible to replace all occurance of enums into enum class via script ([proof of concept](https://github.com/scarf005/Cataclysm-BN/blob/df0e3900b3527d5c779151d28bb8fa0e7b3a2c23/tools/enum_migration/grammar.ebnf))

**Comment by scarf005** [comment]
[proof of concept](https://github.com/scarf005/Cataclysm-BN/blob/df0e3900b3527d5c779151d28bb8fa0e7b3a2c23/tools/enum_migration/grammar.ebnf) script
![image](https://user-images.githubusercontent.com/54838975/167865017-b4ead026-9cf7-4526-9f29-5e2f3709ea64.png)

**Comment by olanti-p** [comment]
Speaking of automation, ReSharper C++ has built-in conversion of C enums to scoped enums, but I don't have a license.
Anyway, Visual Studio provides both warnings for unscoped enums and can list their usages, so it's essentially just some busywork.
